apps/studio/src/server/modules/user/inactiveUsers.service.ts

```diff
@@ -0,0 +1,198 @@
+        if (deletedPermissions.length === 0) return []
+
+        const siteIdsUserHasPermissionsFor = deletedPermissions.map(
+          (p) => p.siteId,
+        )
+
+        return await tx
+          .with("siteAdmins", (eb) =>
+            eb
+              .selectFrom("Site")
+              .leftJoin(
+                "ResourcePermission",
+                "ResourcePermission.siteId",
+                "Site.id",
+              )
+              .innerJoin("User", "User.id", "ResourcePermission.userId")
+              .where("Site.id", "in", siteIdsUserHasPermissionsFor)
+              .where("ResourcePermission.userId", "!=", user.id) // don't want to ask users to ask themselves for permissions
+              .where("ResourcePermission.userId", "not in", userIdsToDeactivate)
+              .where("ResourcePermission.deletedAt", "is", null)
+              .where("ResourcePermission.role", "=", RoleType.Admin) // should only give the admin emails to request reactivation permissions from
+              .where("User.email", "not in", ISOMER_ADMINS_AND_MIGRATORS_EMAILS) // we don't want to send emails to admins and migrators
+              .select([
+                "Site.id as siteId",
+                db.fn
+                  .agg<string[]>("array_agg", ["User.email"])
+                  .as("adminEmails"),
+              ])
+              .groupBy("Site.id"),
+          )
+          // Needed as we still want the site records even if there are no other users with permissions for that site
+          .with("baseSites", (eb) =>
+            eb
+              .selectFrom("Site")
+              .where("Site.id", "in", siteIdsUserHasPermissionsFor)
+              .select(["Site.id as siteId", "Site.name as siteName"]),
+          )
+          .selectFrom("baseSites")
+          .leftJoin("siteAdmins", "siteAdmins.siteId", "baseSites.siteId")
+          .select([
+            "baseSites.siteName",
+            sql`COALESCE("siteAdmins"."adminEmails", ARRAY[]::text[])`.as(
+              "adminEmails",
+            ),
+          ])
+          .execute()
+          .then(
+            (result) =>
+              result as AccountDeactivationEmailTemplateData["sitesAndAdmins"],
+          )
+      })
+  } catch (error) {
+    if (
+      // Handle serializable transaction errors gracefully
+      error instanceof Error &&
+      "code" in error &&
+      error.code === PG_ERROR_CODES.serializationFailure
+    ) {
+      // Return early as the user may have already been deactivated by another concurrent transaction
+      return
+    }
+    // Re-throw other errors
+    throw error
+  }
+
+  // If there are no sites and admins, we don't need to send an email
+  // Note: this is to ensure there's idempotency and not send duplicate emails
+  if (sitesAndAdmins.length === 0) return
+
+  try {
+    await sendAccountDeactivationEmail({
+      recipientEmail: user.email,
+      sitesAndAdmins,
+    })
+  } catch {
```

Comment:
[issue] not a fan of this, because it seems to mix in account sending iwth deactivation. should we let the controller handle this?